compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -360,16 +348,35 @@ public BType checkExpr(BLangExpression expr, SymbolEnv env, BType expType, Diagn
 
         expr.accept(this);
 
-        expr.type = resultType.tag != TypeTags.INTERSECTION ? resultType :
-                ((BIntersectionType) resultType).effectiveType;
+        if (resultType.tag == TypeTags.INTERSECTION) {
+            resultType = ((BIntersectionType) resultType).effectiveType;
+        }
+
+        expr.type = resultType;
         expr.typeChecked = isTypeChecked;
         this.env = prevEnv;
         this.expType = preExpType;
         this.diagCode = preDiagCode;
-        if (resultType.tag != TypeTags.SEMANTIC_ERROR) {
+
+        validateAndSetExprExpectedType(expr);
+
+        return resultType;
+    }
+
+    private void validateAndSetExprExpectedType(BLangExpression expr) {
+        if (resultType.tag == TypeTags.SEMANTIC_ERROR) {
+            return;
+        }
+
+        if (expr.getKind() != NodeKind.RECORD_LITERAL_EXPR ||
+                expr.expectedType == null ||
+                expr.expectedType.tag != TypeTags.MAP ||
+                expr.type.tag != TypeTags.RECORD) {
```

Comment:
Wouldn't the inverse of this check read better?

Reply:
I actually found this more readable :smiley: But when swapping the comment, I changed it to return early if it matches the inverse, so it's kinda swapped now.